llvm/include/llvm/Analysis/DXILMetadataAnalysis.h

```diff
@@ -0,0 +1,85 @@
+//=- DXILMetadataAnalysis.h - Representation of Module metadata --*- C++ -*-=//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_ANALYSIS_DXILMETADATA_H
+#define LLVM_ANALYSIS_DXILMETADATA_H
+
+#include "llvm/ADT/MapVector.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/DXILABI.h"
+#include "llvm/Support/VersionTuple.h"
+#include "llvm/TargetParser/Triple.h"
+
+namespace llvm {
+
+namespace dxil {
+
+struct ModuleMetadataInfo {
+  VersionTuple DXILVersion{};
+  VersionTuple ValidatorVersion{};
+  VersionTuple ShaderModelVersion{};
+  Triple::EnvironmentType ShaderStage = Triple::UnknownEnvironment;
+
+  void init(Module &);
```

Comment:
Rather than having an `init` method in the public interface like this it's better to construct the analysis result in a `static` function or a helper class in an anonymous namespace in the `.cpp` file. That way we don't need to leak implementation details to the header.

Reply:
A constructor is still in the public interface, and the idea is really that we want the initialization to be internal to the analysis. It doesn’t make sense to create this object outside of the analysis, as the analysis is the way to get the analysis result.

Further, if the helper that builds this is static/anonymous, then if we need to make it depend on other analyses or something like that the changes can be contained in the cpp file, which is nice for encapsulation.

Reply:
Changes pushed to use an anonymously scoped function instead of using a constructor.

Reply:
Style nit: llvm usually only uses anonymous namespaces for classes/structs. Local functions are generally just declared “static”.

Reply:
> Style nit: llvm usually only uses anonymous namespaces for classes/structs. Local functions are generally just declared “static”.

Changed to `static` function.